Add MixtureSameFamily distribution

peterroelants opened the pull request:
Add a NumPyro-native implementation of `MixtureSameFamily` to model marginalized mixture distributions, where the components can be given by a vectorized distribution (implies that all the components are from the same family).

This was requested, and deemed relevant, in https://github.com/pyro-ppl/numpyro/issues/1081

I would love to get some feedback on this PR to help get this in.

peterroelants commented:
Updated the PR by adding `mean` and `variance` properties to the mixture based on PyTorch's `[ MixtureSameFamily`](https://github.com/pytorch/pytorch/blob/4753100a3baa96273204c361c8452afb7b59836f/torch/distributions/mixture_same_family.py#L8) class.

Added better integration with existing tests.

peterroelants commented:
One thing I'm still wondering and what I would like to add:
- What would be the best way to be able to access the mixture indices of samples from `.sample()`?
  - Return them in a tuple together with the samples?
  - Create a separate `sample_with_indices` method that returns the `(samples, indices)` as a tuple?
  - Other solution?

fehiepsi commented on test/test_distributions.py:
```diff
@@ -553,6 +608,7 @@ def test_dist_shape(jax_dist, sp_dist, params, prepend_shape):
 
 
+# TODO: prepend_shape not used?
 @pytest.mark.parametrize("prepend_shape", [(), (2,), (2, 3)], ids=str)
```
Good catch! I think we can safely remove this.

peterroelants replied:
Should I remove it in this PR?

fehiepsi replied:
Sure, let's remove it. Thank you!

peterroelants replied:
Removed

fehiepsi reviewed (commented):
Hi @peterroelants, the PR is excellent. Thank you for contributing!

> Create a separate sample_with_indices method that returns the (samples, indices) as a tuple?

This sounds reasonable to me. Currently, we also collapse all latent samples for conjugate distributions. Do you have some applications in mind? With `sample_with_indices`, I guess users will need to write come like
```
def model():
    if data is None:
        data, indices = dist.Mixture(...).sample_with_indices(numpyro.prng_key())
        numpyro.deterministics("indices", indices)
    numpyro.sample("obs", dist.Mixture(...), obs=data)
```

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,210 @@
+            self.mixture_size,
+        )
+        assert (
+            self._component_distribution.batch_shape == expected_component_batch_shape
```
Are you suggesting to only make sure the mixture size is compatible between the component distribution and the mixture distribution? And to have the batch shape being broadcasted?
I could update the code to allow this.

fehiepsi replied:
Yup, I think it is nice to have batch shape being broadcasted.

peterroelants replied:
Updated to only check the mixture_size

peterroelants commented:
> Alternatively, you can implement the default `sample_with_intermediates` method and let `sample` method returns the first output of `sample_with_intermediates`. 

I didn't know about `sample_with_intermediates`. I guess this need to behave similar to the `sample_with_indices` I suggested?

> With this, you need to add a new argument for `log_prob`:
> ```
> def log_prob(self, value, intermediates=None):
>     del intermediates
> ```
Could you explain why this new argument for `log_prob` would be needed?

fehiepsi commented:
Historically, in NumPyro, we don't cache the intermediate values of transformed distributions. So `sample` method needs to return `intermediates` so that when evaluating log probability, we don't have to transform the transformed value back to the base domain. Here the situation is a bit different - the intermediate value `indices` does not help you with evaluating log probability. But it can be nice if we use that API (of transformed distribution).

peterroelants commented:
@fehiepsi Thank you for the review and suggestions. 

I updated the code based on your suggestions and added the `sample_with_intermediates` method. I amended and pushed the commit.

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+            "needs to correspond to the mixture_size={self._mixture_size}!"
+        )
+        super().__init__(
+            batch_shape=self.mixing_distribution.batch_shape,
```
done

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        """Expand given probability array with broadcastable event shape."""
+        broadcastable_shape = (
+            *(s for s in self.batch_shape),
+            self.mixture_size,
```
I personally like these small abstractions like `_expand_probs_event_dim`, which imo are easier to read than the one liner. However, I don't feel strongly about this, so I followed your suggestion.

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        )
+        super().__init__(
+            batch_shape=self.mixing_distribution.batch_shape,
+            event_shape=component_distribution.event_shape,
```
Not sure what exactly you meant so you might want to double check. I changed it to only use the arguments in the checks (instead of object properties).

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+
+    @property
+    def _mixture_dim(self):
+        return len(self.batch_shape)
```
Yes, updated according to suggestion.

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+    def _mixture_dim(self):
+        return len(self.batch_shape)
+
+    def _expand_mixture_dim(self, a):
```
updated according to suggestion.

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        key_comp, key_ind = jax.random.split(key)
+        # Samples from component distribution will have shape:
+        #  (*sample_shape, *batch_shape, mixture_size, *event_shape)
+        samples = self._component_distribution.sample(key_comp, sample_shape)
```
Done both

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        #  (*sample_shape, *batch_shape, mixture_size, *event_shape)
+        samples = self._component_distribution.sample(key_comp, sample_shape)
+        # Sample selection indices from the categorical (shape will be sample_shape)
+        indices = self.mixing_distribution.sample(key_ind, sample_shape)
```
Done

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        n_expand = self.event_dim + 1
+        indices_expanded = indices.reshape(indices.shape + (1,) * n_expand)
+        # Select samples according to indices samples from categorical
+        axis_to_select = -(self.event_dim + 1)
```
You're right. Updated

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        value_broadcastable = value.reshape(
+            (*value.shape[:nb_value_dims], 1, *self.event_shape)
+        )
+        component_log_probs = self._component_distribution.log_prob(value_broadcastable)
```
updated according to suggestion.

peterroelants commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,209 @@
+        nb_value_dims = len(value.shape) - self.event_dim  # Without event dim
+        value_broadcastable = value.reshape(
+            (*value.shape[:nb_value_dims], 1, *self.event_shape)
+        )
```
Updated according to suggestion

peterroelants commented:
@fehiepsi Thanks for the detailed review. I:

- Updated the PR according to your suggested changes
- Added a test for the batch dimension broadcasting
- Rebased on the latest master

fehiepsi commented on numpyro/distributions/mixtures.py:
```diff
@@ -0,0 +1,203 @@
+        )
+        return jnp.sum(cdf_components * self.mixing_distribution.probs, axis=-1)
+
+    def sample_with_intermediates(self, key, sample_shape=()):
```
Could you add docs mention what this method returns?

peterroelants replied:
Added a docstring mentioning the return tuple

fehiepsi reviewed (approved):
Looks great to me! Thanks for adding many tests, @peterroelants! Could you also add an entry to [docs](https://github.com/pyro-ppl/numpyro/blob/master/docs/source/distributions.rst)?

peterroelants commented:
> Could you also add an entry to [docs](https://github.com/pyro-ppl/numpyro/blob/master/docs/source/distributions.rst)?

@fehiepsi I added an entry to the docs, I wasn't sure where to put it, so I created a new section called "Mixture Distributions" just before "Directional Distributions". I can imagine more mixture distributions being added in the future

fehiepsi commented:
> I can imagine more mixture distributions being added in the future

I hope so. :D Currently, I only know about GMM. Waiting for more interesting distributions to come.

fehiepsi reviewed (approved):
The PR looks great to me. Thanks, @peterroelants!

peterroelants commented:
@fehiepsi thanks for the review!

Do I need to do anything else to get this merged in? (I don't have the right permissions myself to merge PRs)

fehiepsi commented:
Sorry, I just waited for CI finish its job. :)
